Ability to set HOMEPAGE_COURSE_MAX via SiteConfiguration.

mtyaka opened the pull request:
This patch makes it possible to set `HOMEPAGE_COURSE_MAX` via `SiteConfiguration` in the django admin. The setting controls the number of courses displayed on the homepage.

**JIRA ticket**: https://openedx.atlassian.net/browse/OSPR-1817

**Dependencies**: None

**Sandbox URL**: 
- LMS: https://pr15476.sandbox.opencraft.hosting/
- Studio: https://studio-pr15476.sandbox.opencraft.hosting/

The sandbox is set up with 4 courses and `HOMEPAGE_COURSE_MAX` is set to 2 in site configurations.

**Partner information**: 3rd party-hosted open edX instance

**Merge deadline**: None

**Testing instructions**:

1. Create and publish several courses in the Studio, so that they are visible on the LMS homepage.
1. By default there is no limit to the number of courses displayed on the homepage, so you should see all of the courses you added.
1. Go to the Site Configurations section of the django admin (/admin/site_configuration/siteconfiguration/).
1. Set up your Site/Site configuration and add `"HOMEPAGE_COURSE_MAX": 2` (or any other number you want to test).
1. Observe that the LMS homepage now only lists the number of pages you set in the configuration, and it shows a "View all Courses" link, which takes you to a page that lists all courses.

**Note: the homepage is cached and you may have to wait up to three minutes for the cache to update. If you don't want to wait, you can `rm -r /tmp/mako_*` and/or `sudo service memcached restart` before you see the changes. You may even have to restart edxapp (`/edx/bin/supervisor restart edxapp:*`).**

**Reviewers**
- [x] @pomegranited 
- [ ] edX reviewer[s] TBD

edx-webhook commented:
Thanks for the pull request, @mtyaka! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-platform&number=15476"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

pomegranited commented on lms/djangoapps/courseware/tests/test_microsites.py:
```diff
@@ -129,6 +133,26 @@ def test_no_configuration_anonymous_homepage_content(self):
+
+        # Site now has three visible courses, but the HOMEPAGE_COURES_MAX setting is set to 2,
+        # so we should see a link to all courses.
+        self.assertContains(resp, 'View all Courses')
```
Can you also please test that the number of courses displayed matches the `HOMEPAGE_COURES_MAX` setting?

pomegranited replied:
Could you also test the case where HOMEPAGE_COURES_MAX exceeds the number of courses?

mtyaka commented:
Squashed from 68e15aa8d72a98b19e0cf8f00b4d218dc7fe3a1c

mtyaka commented:
@pomegranited Thanks a lot for the review, for locating the hardcoded cache timeout, and for pinging the doc team.

mtyaka commented:
jenkins run bokchoy

edx-webhook commented:
Thanks for the pull request, @mtyaka! I've created [OSPR-1817](https://openedx.atlassian.net/browse/OSPR-1817) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

edx-webhook commented:
Thanks for the pull request, @mtyaka! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-platform&number=15476"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

edx-webhook commented:
Thanks for the pull request, @mtyaka! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-platform&number=15476"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

pomegranited commented:
@nedbat Could you let us know when this PR could be reviewed?  It's a small one but a goodie, and would be great to get merged.

nedbat commented:
@edx/learner Who can take this on?

robrap commented on common/djangoapps/student/views.py:
```diff
@@ -195,6 +195,11 @@ def index(request, extra_context=None, user=AnonymousUser()):
 
+    # Maximum number of courses to display on the homepage.
+    context['homepage_course_max'] = configuration_helpers.get_value(
+        'HOMEPAGE_COURSE_MAX', settings.HOMEPAGE_COURSE_MAX
```
Is there already a test for the original `settings.HOMEPAGE_COURSE_MAX`?  Do you want a test to ensure that stays the default?

mtyaka replied:
Good idea, thanks. I added a couple of extra tests.

mtyaka commented:
Thanks for reviewing into this @robrap! I added extra tests to ensure `settings.HOMEPAGE_COURSE_MAX` is used if site configuration variable is not set. Also rebased on top of master.

robrap commented:
Thanks for the update @mtyaka.  I just saw the testing comment from @pomegranited:
> Also tested the case where HOMEPAGE_COURSE_MAX exceeds the number of available courses.

Seems like a simple and useful test addition.  Let me know your thoughts.  Thanks.

mtyaka commented:
jenkins run a11y

mtyaka commented:
> Seems like a simple and useful test addition. Let me know your thoughts. Thanks.

Makes sense @robrap, test added, thanks.

mtyaka commented:
Thanks @robrap!

Squashed and rebased from dbadcd97cd8d4713eef7773c28f088b3bf64194a.

mtyaka commented:
jenkins run a11y

mtyaka commented:
jenkins run quality

edx-webhook commented:
@mtyaka 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/15476&entry.1484655318=https://github.com/mtyaka&entry.752974735=2017-07-06+08:52&entry.1917517419=2017-11-08+17:21&entry.2133058324=Yes) so we can improve your experience in the future.
